aspnetcore/fundamentals/app-state.md

```diff
@@ -193,88 +279,111 @@ app.Use(async (context, next) =>
 });
 ```
 
-Later in the pipeline, another middleware could access it:
+Later in the pipeline, another middleware can access the value of `isVerified`:
 
 ```csharp
 app.Run(async (context) =>
 {
-    await context.Response.WriteAsync("Verified request? " +
-        context.Items["isVerified"]);
+    await context.Response.WriteAsync($"Verified: {context.Items["isVerified"]}");
 });
 ```
 
-For middleware that will only be used by a single app, `string` keys are acceptable. However, middleware that will be shared between applications should use unique object keys to avoid any chance of key collisions. If you are developing middleware that must work across multiple applications, use a unique object key defined in your middleware class as shown below:
+For middleware that's only used by a single app, `string` keys are acceptable. Middleware shared between app instances should use unique object keys to avoid key collisions. The following example shows how to use a unique object key defined in a middleware class:
 
-```csharp
-public class SampleMiddleware
-{
-    public static readonly object SampleKey = new Object();
+::: moniker range=">= aspnetcore-2.0"
 
-    public async Task Invoke(HttpContext httpContext)
-    {
-        httpContext.Items[SampleKey] = "some value";
-        // additional code omitted
-    }
-}
-```
+[!code-csharp[](app-state/samples/2.x/SessionSample/Middleware/HttpContextItemsMiddleware.cs?name=snippet1&highlight=4,13)]
+
+::: moniker-end
+
+::: moniker range="< aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/1.x/SessionSample/Middleware/HttpContextItemsMiddleware.cs?name=snippet1&highlight=5,14)]
+
+::: moniker-end
 
 Other code can access the value stored in `HttpContext.Items` using the key exposed by the middleware class:
 
-```csharp
-public class HomeController : Controller
-{
-    public IActionResult Index()
-    {
-        string value = HttpContext.Items[SampleMiddleware.SampleKey];
-    }
-}
-```
+::: moniker range=">= aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/2.x/SessionSample/Pages/Index.cshtml.cs?name=snippet3)]
 
-This approach also has the advantage of eliminating repetition of "magic strings" in multiple places in the code.
+::: moniker-end
 
-## Application state data
+::: moniker range="< aspnetcore-2.0"
+
+[!code-csharp[](app-state/samples/1.x/SessionSample/Controllers/HomeController.cs?name=snippet3)]
+
+::: moniker-end
+
+This approach also has the advantage of eliminating the use of key strings in the code.
+
+## Cache
+
+Caching is an efficient way to store and retrieve data. The app can control the lifetime of cached items. For more information, see the [Cache responses](xref:performance/caching/index) topic.
```

Comment:
Cached data is not associated with an specific request, user, or session. Be careful not to cache user specific information that may be retrieved by other users / requests.